### PR TITLE
protocols/kad: Fix right shift overflow panic in record_received 

### PR DESCRIPTION
**Pull request by mxinden**
Within `Behaviour::record_received` the exponentially decreasing
expiration based on the distance to the target for a record is
calculated as following:

1. Calculate the amount of nodes between us and the record key beyond
the k replication constant as `n`.

2. Shift the configured record time-to-live `n` times to the right to
calculate an exponentially decreasing expiration.

The configured record time-to-live is a u64. If `n` is larger or equal
to 64 the right shift will lead to an overflow which panics in debug
mode.

This patch uses a checked right shift instead, defaulting to 0 (`now +
0`) for the expiration on overflow.

Fixes https://github.com/libp2p/rust-libp2p/issues/1491.

### CODE REVIEW

File: protocols/kad/src/behaviour/test.rs [hunk 1]
```diff
@@ -682,3 +684,44 @@ fn exceed_jobs_max_queries() {
         })
     )
 }
+
+#[test]
```
Comment (twittner) [inline protocols/kad/src/behaviour/test.rs]: I think in libp2p the convention is to put attributes below comments.

File: protocols/kad/src/handler.rs [hunk 2]
```diff
@@ -364,6 +364,15 @@ pub struct KademliaRequestId {
     connec_unique_id: UniqueConnecId,
 }
 
+#[cfg(test)]
+impl Default for KademliaRequestId {
```
Comment (twittner) [inline protocols/kad/src/handler.rs]: As long as `KademliaRequestId` implements `Clone`, this `Default` impl might as well be available all the time.

Comment (tomaka) [reply protocols/kad/src/handler.rs]: This is an opaque implementation detail. Implementing `Default` on it doesn't make any sense. Maybe it could contain a channel or an `Arc` in the future.

File: protocols/kad/src/behaviour/test.rs [hunk 3]
```diff
@@ -682,3 +684,44 @@ fn exceed_jobs_max_queries() {
...
+    }
+
+    let connection_id = ConnectionId::new(1234);
+    let request_id = KademliaRequestId::default();
+    let record = Record::new(
+        record::Key::from(target.clone().into_bytes()),
+        vec![],
+    );
+
+    kad.record_received(target, connection_id, request_id, record);
```
Comment (tomaka) [inline protocols/kad/src/behaviour/test.rs]: We shouldn't be testing private implementation details. The parameter of this method are totally arbitrary. The fact that it's a separate method is also totally arbitrary.

File: protocols/kad/src/handler.rs (hunk 2 above)
Comment (twittner) [reply protocols/kad/src/handler.rs]: I understood `KademliaRequestId` to be some sort of unique one-time token and `Clone` defeats that purpose but I did not look much into it so feel free to ignore the criticism.

File: protocols/kad/src/behaviour.rs [hunk 4]
```diff
@@ -957,7 +957,7 @@ where
         let k = self.queries.config().replication_factor.get();
         let num_beyond_k = (usize::max(k, num_between) - k) as u32;
         let expiration = self.record_ttl.map(|ttl|
-            now + Duration::from_secs(ttl.as_secs() >> num_beyond_k)
+            now + Duration::from_secs(ttl.as_secs().checked_shr(num_beyond_k).unwrap_or(0))
```
Comment (arkpar) [inline protocols/kad/src/behaviour.rs]: Does it make sense to put records that are immediately expired in the store on line 987

File: protocols/kad/src/handler.rs (hunk 2 above)
Comment (mxinden) [reply protocols/kad/src/handler.rs]: I am fine either way. I can't come up with a use-case for `Default` outside of testing though.

File: protocols/kad/src/behaviour/test.rs (hunk 3 above)
Comment (mxinden) [reply protocols/kad/src/behaviour/test.rs]: I understand your concern. What are you advocating for instead: (1) remove the test or (2) write a larger test that reproduces the same behaviour?

> All we're doing with this test is locking the current implementation as it is.

I don't think a test should ever _lock_ an implementation. While I am in favor of having a test for each regression, I am also in favor of removing outdated tests if necessary.

File: protocols/kad/src/behaviour.rs (hunk 4 above)
Comment (mxinden) [reply protocols/kad/src/behaviour.rs]: As far as I can tell the record would be removed by the next `PutRecordJob::poll` invocation. Maybe @romanb has a better answer.

Never the less I would advocate for addressing this in a follow up pull-request.

File: protocols/kad/src/handler.rs (hunk 2 above)
Comment (romanb) [reply protocols/kad/src/handler.rs]: > I understood `KademliaRequestId` to be some sort of unique one-time token and `Clone` defeats that purpose but I did not look much into it so feel free to ignore the criticism.

Just to give some context: `KademliaRequestId` did not use to be `Clone` but it had to be as a result of https://github.com/libp2p/rust-libp2p/pull/1440 requiring `Clone` for the `TInEvent` of any handler, since it is now possible to "broadcast" a single such event to multiple handlers (which in turn was a requirement discovered during the substrate integration in https://github.com/paritytech/substrate/pull/5066) and `KademliaRequestId` is part of these events in `libp2p-kad`. It would of course be better to not have that `Clone`, but I didn't come up with a way to do so.

File: protocols/kad/src/behaviour.rs (hunk 4 above)
Comment (romanb) [reply protocols/kad/src/behaviour.rs]: Yes, expired records are regularly purged (and a `RecordStore` may of course just drop expired records at any time anyway), but it may nevertheless be a good optimisation that may also make it a bit more difficult to somewhat maliciously or accidentally (temporarily) fill a node's store with already expired records. @mxinden if you could create an issue for it, that would be great, thanks!

Comment (mxinden) [reply protocols/kad/src/behaviour.rs]: See https://github.com/libp2p/rust-libp2p/pull/1496.

**Comment by romanb** [comment]
Thanks for fixing this, I wasn't even aware bit shifts could overflow in this way. A regression test is a good idea but I don't see why it needs to be so complicated. Why not extract the code that calculates the expiry into a standalone function and writing small property and or / regresstion test for that? Even if you want the function to actually take a `KBucketsTable` as argument, so that it uses `count_nodes_between` internally, there shouldn't be a need for such a probabilistic generation of so many peer IDs, since `Key::for_distance` can be used in combination with the `ClosestBucketsIter` to deterministically fill the buckets as desired ([this test](https://github.com/libp2p/rust-libp2p/blob/master/protocols/kad/src/kbucket.rs#L641-L661) may give inspiration). In other words, when testing a particular edge case requires seemingly too much surrounding boilerplate in order to test what you want, it is often a good idea to partition the code further so that the part you're interested in testing in isolation is easier to test, without so much surrounding ceremony. This seems to be such a case.

**Comment by mxinden** [comment]
@romanb thanks for the input. Can you take another look?

File: protocols/kad/src/behaviour.rs [hunk 5]
```diff
@@ -946,8 +946,6 @@ where
             return
         }
 
-        let now = Instant::now();
```
Comment (romanb) [inline protocols/kad/src/behaviour.rs]: I think it is preferable to leave this here as it can be reused multiple times (cf. https://github.com/libp2p/rust-libp2p/pull/1496). And since `Instant::now()` is a side-effecting function for obtaining some notion of system time, usually both testability and performance is improved by using it mostly in top-level functions and passing the instant down to others where needed (see also the other comments).

**Comment by mxinden** [comment]
Suggested simplifications look good to me. @romanb would you mind taking another look?
